### DISCUSSION TITLE
can nvm install node from local node exe file

### DISCUSSION THREAD

**Discussion by kulkarniash-source** [coreybutler/nvm-windows#695] [Support] [answered]
Hi
nvm is a great tool, exactly what i am looking for, but i work in an environment we cannot download software from internet, but only pre approved software, 
i have downloaded nodejs 12 and nodejs 15 exe files for windows 10, can i configure nvm to install these files instead of going out to internet and download 
something like node_mirror to point towards local folder

**Comment by coreybutler** [comment, marked as answer]
Currently, local installation sources are not supported. NVM4W relies on remote servers hosting Node.

There is a workaround though. You could mimic what NVM4W does during the install process. In the nvm installation root, create a folder called `v12.x.x` and `v15.x.x` (use the specific versions instead of the x's). Inside this directory, paste the `node.exe` file. You'll most likely want npm, npm.cmd, npx, npx.cmd, and possibly nodevars.bat. You'll also need to create `node_modules` in this directory and place any npm modules there (as well as npm). If you have global npm modules, they often have command files (i.e. `module` and `module.cmd`). You'll have to do this for each version. Once that is done, you can `nvm use x.x.x` to switch versions.

Thats' a fair amount of work. It may be easier to use a computer that can install over the internet to install those two versions, then use a USB drive to copy the file structure onto the computer that cannot access the internet/downloads.

This type of situation has come up once before, many years ago, with a school district that did not have very reliable internet. At the time, I explored building a download server for running locally inside a trusted network. I drew a few conclusions. 

First, everyone could potentially benefit from local download servers. Bandwidth adds up quickly for organizations who can access the internet. A local download server/cache can reduce bandwidth. Organizations who want more control over the download process would obviously benefit from this. This includes organizations who wish to use customized versions of Node (such as Nodesource's supported releases).

Second, a download server is beyond the scope of NVM4W. It's really a separate tool for managing work environments. I have been thinking about releasing an affordable solution like this for years. I'm seriously considering building this service for rt (NVM4W's successor - see the welcome message in the discussions for details).

The idea of a local `node_mirror` directory is a good one - worth exploring further. I won't add this to NVM4W, but I'll experiment with the idea for rt.

I'd be interested in feedback from folks who have similar restrictions in their environment. A local install source seems like the simplest solution, but it seems like there could be administrative challenges in maintaining an environment relying only on the file system.
